### ISSUE TITLE
Move away from using `view` helper

### ISSUE THREAD

**Issue by wvteijlingen** [piceaTech/ember-rapid-forms#39]
The `{{view}}` helper will be removed in Ember 2.0, and is already deprecated in 1.13. It seems the architecture of this library is built upon using `{{view}}` to load the `controlView`. I don't see any quick-fixes for this, other than to change the architecture. Unfortunately, it seems the `{{component}}` helper only takes a component name as a string.

Has there been any thought about this already? Otherwise I'll need to start looking into other libraries or we need to do a big rewrite for this one or the original ember-forms.

**Comment by pedrokiefer** [comment]
I'm all in for a big rewrite. The way we pass attributes from a parent component to its child is just horrible to say the least. If we could compose the component just using javascript, that would be the right way to go, imho. I think this would also make it easier to add custom widgets to a form.

@wvteijlingen do you have any suggestion on how we should handle this new architecture?

**Comment by wvteijlingen** [comment]
Honestly, I think https://github.com/Emerson/ember-form-master-2000 already has a better basic architecture. It does miss some niceties like delayed validation, enabling of the submit button, etc. though.

But I don't know whether (re)writing this form library is worth the investment at this point. I would love for [this](https://github.com/emberjs/rfcs/blob/contextual-component-lookup/text/0000-contextual-component-lookup.md) RFC to go though. That will open the door to some real nice Rails-like form building.

**Comment by pedrokiefer** [comment]
The RFC is what I always thought forms should be handled!

Just found out that there is a `{{#component}}`  helper, not sure if it can help us while the RFC is not out.

**Comment by BenjaminHorn** [comment]
@pedrokiefer  could you please let us know, how the new architecture will look like? Some pseudo code will be great.

**Comment by pedrokiefer** [comment]
I really like the architecture proposed on the RFC. 

**DISCLAIMER: I've just made this up, and I'm totally open to discussions!**

So we would get something like this on the templates:
```hbs
{{#my-form-for model as |f|}}
  {{f.fullName}}
  {{f.password}}
  {{f.date}}
{{/my-form-for}}
```

And something like this defining your custom form:
```js
// app/component/my-form-for.js
import Ember from "ember";
import RapidFormMixin from "ember-rapid-forms";
import createInput from "ember-rapid-forms";
import DateInput from '../date-input'; 

export default Ember.Component.extend(RapidFormMixin, {
   formHelpers: {
      fullName: createInput('text', { property: 'fullName', label: 'Full Name' }),
      password: createInput('password', { property: 'password', label: 'Password'}),
      date: createInput(DateInput, { property: 'date' })
   }
});
```

Not sure on how a custom component would be created, if using a mixin extending a current component, or something like a wrapper object.

What do you think about it @spruce, @wvteijlingen, @BenjaminHorn ?

**Comment by BenjaminHorn** [comment]
@pedrokiefer  
Very interesting approach. However it feels a little bit odd, not to define the whole form in a template.
I think I like more the classical approach.

```javascript
// app/components/my-form.js
{{#my-form model=model}}
  {{!-- ships with the addon --}}
  {{control-group type="my-input" property="name" }}
  {{!-- some other components --}}
  {{control-group type="masked-input" property="phone" mask="9-999-999" }}
  {{control-group type="select-2" content="genders"  property="gender" optionValuePath="id"}}
{{/my-form}}
```
the control-group template could be something like this
```javascript
// app/templates/components/control-group,hbs
{{#if error}}
  {{error}}
{{/error}}
{{component type value=value ...}}
```
But same issue: no idea how to pass individual attributes (like `mask` and `optionValuePath`) though control-group.

**Comment by wvteijlingen** [comment]
I think the approach in the RFC is actually the best approach at this point. I think it's more intuitive to call the input components as `{{f.input "fullName"}}` instead of `{{f.fullName}}`. Custom components would just be subclasses that mixin the `RapidFormMixin` then.

**Comment by pedrokiefer** [comment]
I really like the way [django-crispy-forms](http://django-crispy-forms.readthedocs.org/en/latest/index.html) work. On the template you just have `{% crispy form %}`, and everything else on how this form will be rendered are defined on the form class, either statically or dynamically. Going on with this idea, we would simply have on the template:

```hbs
{{my-form-for currentModel}}
``` 

And all the behaviour of the form would be on the `JS` side. You can even have different templates for the same form, ie. `bootstrap3`, `foundation`, etc.

**Comment by BenjaminHorn** [comment]
@pedrokiefer  I know squadoosh about django or crispy form, but I like the idea, to have **all** the definitions are at one place. (I still think it would be better place for it in the template, but then we could not use 3rd party components as flexible as I would like to see https://github.com/piceaTech/ember-rapid-forms/issues/39#issuecomment-115311828).

**Comment by wvteijlingen** [comment]
> And all the behaviour of the form would be on the JS side.

But the architecture you proposed doesn't really do this. Depending on what you mean with "behavior". It still leaves the layout of the form to the template.

I think if you want to have one component for an entire form, you could just create that component and in the template define the fields using the `{{f.input}}` components.

Having a generic input component like this also allow you to create simple logo less forms without the need to write any JS.

**Comment by BenjaminHorn** [comment]
>It still leaves the layout of the form to the template.

@wvteijlingen  I don't know how @pedrokiefer wants to do this, but from http://django-crispy-forms.readthedocs.org/en/latest/layouts.html#fundamentals

>Django-crispy-forms defines another powerful class called Layout, which allows you to change the way the form fields are rendered. This allows you to set the order of the fields, wrap them in divs or other structures, add html, set ids, classes or attributes to whatever you want, etc. And all that without writing a custom form template, using programmatic layouts. Just attach the layout to a helper, layouts are optional, but probably the most powerful thing django-crispy-forms has to offer.

**Comment by spruce** [comment]
If you are still interested in using this project. I just finished writing a massive PR which fixes all the deprecations of ember 1.13. All the tests pass on 2.0, 2.1 and 2.2 canary. So it is now future proof

**Comment by GCorbel** [comment]
@spruce it seems to work for me.
